text/0000-guaranteed-tco.md

```diff
@@ -0,0 +1,608 @@
+    drop(b);
+    become y(_tmp);
+}
+```
+
+
+This early dropping allows the compiler to avoid many complexities associated with deciding if a call can be TCO,
+instead the heavy lifting is done by the borrow checker and a lifetime error will be produced if references to local
+variables are passed to the called function. To be clear a reference to a local variable could be passed if instead of
+`become` the call would be done with `return y(a);` (or equivalently `y(a)`), indeed this difference between the
+handling of local variables is also the main difference between `return` and `become`.
+
+### Use Case 1: Recursive Algorithm
+As a possible use case let us take a look at creating the sum over a `Vec`. Admittedly an unusual example for Rust as
+this is usually done with iteration. Though, this is kind of the point, without TCO this example can overflow the stack.
+
+```rust
+fn sum_list(data: Vec<u64>, mut offset: usize, mut accum: u64) -> u64 {
+    if offset < data.len() {
+        accum += data[offset];
+        offset += 1;
+        become sum_list(data, offset, accum); // <- become here
+    } else {
+        // Note that this would be a `return accum;`
+        accum
+    }
+}
+```
+
+
+### Use Case 2: Interpreter
+In an interpreter the usual loop is to get an instruction, match on that instruction to find the corresponding function, **call** that function, and finally return to the loop to get the next instruction. (This is a simplified example.)
+
+```rust
+fn exec_instruction(mut self) {
+    loop {
+        let next_instruction = self.read_instr(); // this call can be inlined
+        match next_instruction {
+            Instruction::Foo => self.execute_instruction_foo(),
+            Instruction::Bar => self.execute_instruction_bar(),
+        }
+    }
+}
+```
+
+This example can be turned into the following code, which no longer does any calls and instead just uses jump instructions. (Note that this example might not be the optimal way to use `become`.)
+
+```rust
+fn execute_instruction_foo(mut self) {
+    // foo things ...
+
+    become self.next_instruction();
+}
+
+fn execute_instruction_bar(mut self) {
+    // bar things ...
+
+    become self.next_instruction();
+}
+
+fn next_instruction(mut self) {
+    let next_instruction = self.read_instr(); // this call can be inlined
+    match next_instruction {
+        Instruction::Foo => become self.execute_instruction_foo(),
+        Instruction::Bar => become self.execute_instruction_bar(),
+    }
+}
+```
+
+### Omission of the `become` keyword causes the call to be `return` instead.
+([original example](https://github.com/rust-lang/rfcs/pull/1888#issuecomment-278988088))
+
+This is a potential source of confusion, indeed in a function language where every call is expected to be TCO this would be quite unexpected. (Maybe in functions that use `become` a lint should be applied that enforces usage of either `return` or `become` in functions where at least one `become` is used.)
```

Comment:
I can already depict a very useful `clippy` lint like `#[clippy(enforce_tail_call)]` that can be applied to a function declaration to enforce that all calls in tail position are flagged as proper tail calls and warn otherwise.

Reply:
I kind of like the idea to have `rustc` warn about this in functions where at least one `become` is used.

Reply:
The main question is: Are there actual use cases where some branches in a function lead to a tail call and other branches lead to normal calls?

Reply:
I can't really think of one on the spot. Though you would use `return` not only to call further functions but also to actually return a value. So another question is, should this lint be applied only to call expressions or also implicit returns at the end of a function? If only call expressions, what if those are part of larger expressions?

Reply:
added this to unresolved questions: https://github.com/rust-lang/rfcs/pull/3407/commits/c55c189c44605b9568daf7cd3da661b8abd111f2

resolving this for now

Reply:
i can think of a use-case:
```rust
fn interpreter_instruction_read_byte(state: &mut State) -> Result<(), Box<dyn Error>> {
    match state.inp.bytes().next() {
        Some(Ok(b)) => {
            state.push(b as u64);
            become state.run_next_instruction();
        }
        Some(Err(e)) => {
            // we specifically want a call to a #[inline(never)] function so this frame shows up in the backtrace
            return add_backtrace_to_error(e);
        }
        None => {
            state.push(-1 as u64);
            become state.run_next_instruction();
        }
    }
}
```

Reply:
Also, with the same-argument restriction you really may be forced not to `become` all of your tail calls. I think a lint would not be appropriate unless you can determine that it is both applicable and would result in strictly less work, and even then I would expect some examples like @programmerjake's which would probably use `allow(..)` (assuming that `add_backtrace_to_error()` had the same calling convention such that a `become` would be applicable).